playpen/repo_level_awareness/codeplan.py

```diff
@@ -112,127 +115,203 @@ def __init__(
 
         self._validators_are_stale = False
-
-        return validation_errors
+        logger.debug("Validators are up to date.")
+        return validation_tasks
 
     def get_next_task(self) -> Generator[Task, Any, None]:
-        validation_errors: list[tuple[type, Task]] = []
-        ignored_tasks = []
-
-        prior_error = None
-        # Check to see if validators are stale. If so, run them
-        while True:
-            if self._validators_are_stale:
-                validation_errors = self.run_validators()
-
-            # pop an error of the stack of errors
-            if len(validation_errors) > 0:
-                print(validation_errors.__len__())
-                err = validation_errors.pop(0)
-                if prior_error and fuzzy_equals(prior_error, err[1], offset=2):
-                    ignored_tasks.append(err[1])
-                    continue
-                if any([fuzzy_equals(err[1], v, offset=2) for v in ignored_tasks]):
-                    print(f"failed to solve error, skipping for now\n{err[1]}")
-                    continue
-                yield err[1]  # TODO: This is a placeholder
-                prior_error = err[1]
-                continue
-
-            if len(ignored_tasks) > 0:
-                # Time to give these errors another try since we've fixed everything else
-                ignored_tasks = []
+        prior_task = None
+        self.initialize_task_stacks()
+        while any(self.task_stacks.values()):
+            task = self.pop_task_from_highest_priority()
+            logger.debug("Popped task from stack: %s", task)
+            if self.should_skip_task(task):
+                logger.debug("Skipping task: %s", task)
                 continue
 
-            # if len(self.unprocessed_files) > 0:
-            #     yield Task(self.unprocessed_files.pop(0))
-            #     continue
-
-            break
+            logger.info("Yielding task: %s", task)
+            yield task
+            prior_task = task
+            self.handle_new_tasks_after_processing(task, prior_task)
+
+    def initialize_task_stacks(self):
+        logger.info("Initializing task stacks.")
+        if self._validators_are_stale or not any(self.task_stacks.values()):
+            new_tasks = self.run_validators()
+            logger.debug("New tasks from validators: %s", new_tasks)
+            for task in new_tasks:
+                self.add_task_to_stack(task)
+                logger.debug("Task %s added to stack.", task)
+
+    def add_task_to_stack(self, task: Task):
+        priority = task.priority
+        if priority not in self.task_stacks:
+            self.task_stacks[priority] = []
+            logger.debug("Created new task stack for priority %s.", priority)
+        self.task_stacks[priority].append(task)
+        logger.debug("Task %s added to priority %s stack.", task, priority)
+
+    def pop_task_from_highest_priority(self) -> Task:
```

Comment:
Small performance thing, but since `Task` has `__lt__`, you should be able to use [Python's built-in priority queue stuff](https://docs.python.org/3/library/heapq.html), if you want.

Reply:
Yeah I considered it but thought it was pretty easy to manage myself, but it still could be worth moving to the implementation for terseness/performance. Hopefully the queue doesn't grow that large...